File: android/src/main/kotlin/com/appcues/flutter/AppcuesFlutterPlugin.kt
package com.appcues.flutter

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.net.Uri
import androidx.annotation.NonNull
import com.appcues.AnalyticType
import com.appcues.AnalyticsListener
import com.appcues.Appcues
import com.appcues.AppcuesFrameView
import com.appcues.LoggingLevel

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.embedding.engine.renderer.FlutterRenderer
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

/** AppcuesFlutterPlugin */
class AppcuesFlutterPlugin : FlutterPlugin, MethodCallHandler, ActivityAware {

    private lateinit var methodChannel: MethodChannel
    private lateinit var analyticsChannel: EventChannel
    private lateinit var context: Context

    private lateinit var _renderer: FlutterRenderer
    internal val renderer: FlutterRenderer
        get() = _renderer

    private var _activity: Activity? = null
    internal val activity: Activity?
        get() = _activity

    // this implies the consumer must call "initialize" first, or they will get an error below
    private lateinit var implementation: Appcues

    private val elementTargeting = FlutterElementTargetingStrategy(this)

    private val mainScope = CoroutineScope(Dispatchers.Main)

    override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        context = flutterPluginBinding.applicationContext
        methodChannel = MethodChannel(flutterPluginBinding.binaryMessenger, "appcues_flutter")
        methodChannel.setMethodCallHandler(this)
        analyticsChannel = EventChannel(flutterPluginBinding.binaryMessenger, "appcues_analytics")
        _renderer = flutterPluginBinding.flutterEngine.renderer

        flutterPluginBinding.platformViewRegistry.registerViewFactory(
            "AppcuesFrameView",
            AppcuesFrameViewFactory(this, flutterPluginBinding.binaryMessenger)
        )
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        this._activity = binding.activity
    }

    override fun onDetachedFromActivityForConfigChanges() {
        this._activity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        this._activity = binding.activity
    }

    override fun onDetachedFromActivity() {
        this._activity = null
    }

    override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
        when (call.method) {
            "initialize" -> {
                val accountId = call.argument<String>("accountId")
                val applicationId = call.argument<String>("applicationId")
                if (accountId != null && applicationId != null) {
                    implementation = Appcues(context, accountId, applicationId) {
                        call.argument<HashMap<String, Any?>>("options")?.let {
                            val logging = it["logging"] as? Boolean
                            if (logging != null) {
                                this.loggingLevel = if (logging) LoggingLevel.INFO else LoggingLevel.NONE
                            }

                            val apiHost = it["apiHost"] as? String
                            if (apiHost != null) {
                                this.apiBasePath = apiHost
                            }

                            val settingsHost = it["settingsHost"] as? String
                            if (settingsHost != null) {
                                this.apiSettingsPath = settingsHost
                            }

                            val sessionTimeout = it["sessionTimeout"] as? Double
                            if (sessionTimeout != null) {
                                this.sessionTimeout = sessionTimeout.toInt()
                            }

                            val activityStorageMaxSize = it["activityStorageMaxSize"] as? Double
                            if (activityStorageMaxSize != null) {
                                this.activityStorageMaxSize = activityStorageMaxSize.toInt()
                            }

                            val activityStorageMaxAge = it["activityStorageMaxAge"] as? Double
                            if (activityStorageMaxAge != null) {
                                this.activityStorageMaxAge = activityStorageMaxAge.toInt()
                            }
                        }
                        call.argument<HashMap<String, Any>>("additionalAutoProperties")?.let {
                            this.additionalAutoProperties = it
                        }
                    }
                    analyticsChannel.setStreamHandler(object: EventChannel.StreamHandler {
                        override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
                            implementation.analyticsListener = object: AnalyticsListener {
                                override fun trackedAnalytic(
                                    type: AnalyticType,
                                    value: String?,
                                    properties: Map<String, Any>?,
                                    isInternal: Boolean
                                ) {
                                    events?.success(
                                        hashMapOf(
                                            "analytic" to type.name,
                                            "value" to (value ?: ""),
                                            "properties" to (properties ?: mapOf()),
                                            "isInternal" to isInternal,
                                        )
                                    )
                                }
                            }
                        }

                        override fun onCancel(arguments: Any?) {
                            implementation.analyticsListener = null
                        }

                    })
                    Appcues.elementTargeting = elementTargeting
                    result.success(null)
                } else {
                    result.badArgs("accountId, applicationId")
                }
            }
            "identify" -> {
                val userId = call.argument<String>("userId")
                if (userId != null) {
                    val properties = call.argument<HashMap<String, Any>>("properties");
                    implementation.identify(userId, properties)
                    result.success(null)
                } else {
                    result.badArgs("userId")
                }
            }
            "group" -> {
                val groupId = call.argument<String>("groupId")
                val properties = call.argument<HashMap<String, Any>>("properties");
                implementation.group(groupId, properties)
                result.success(null)
            }
            "track" -> {
                val eventName = call.argument<String>("name")
                if (eventName != null) {
                    val properties = call.argument<HashMap<String, Any>>("properties");
                    implementation.track(eventName, properties)
                    result.success(null)
                } else {
                    result.badArgs("name")
                }
            }
            "screen" -> {
                val title = call.argument<String>("title")
                if (title != null) {
                    val properties = call.argument<HashMap<String, Any>>("properties");
                    implementation.screen(title, properties)
                    result.success(null)
                } else {
                    result.badArgs("title")
                }
            }
            "anonymous" -> {
                implementation.anonymous()
                result.success(null)
            }
            "reset" -> {
                implementation.reset()
                result.success(null)
            }
            "version" -> {
                result.success(implementation.version)
            }
            "debug" -> {
                activity?.let {
                    implementation.debug(it)
                }
                result.success(null)
            }
            "show" -> {
                val experienceId = call.argument<String>("experienceId")
                if (experienceId != null) {
                    mainScope.launch {
                        val success = implementation.show(experienceId)
                        if (success) {
                            result.success(null)
                        } else {
                            result.error("show-experience-failure", "unable to show experience $experienceId", null)
                        }
                    }
                } else {
                    result.badArgs("experienceId")
                }
            }
            "didHandleURL" -> {
                val url = call.argument<String>("url")
                if (url != null) {
                    val activity = this.activity
                    if (activity != null) {
                        val uri = Uri.parse(url)
                        val intent = Intent(Intent.ACTION_VIEW)
                        intent.data = uri
                        result.success(implementation.onNewIntent(activity, intent))
                    } else {
                        result.error("no-activity", "unable to handle the URL, no current running Activity found", null)
                    }
                } else {
                    result.badArgs("url")
                }
            }
            "setTargetElements" -> {
                var viewElements = call.argument<List<HashMap<String, Any>>>("viewElements");
                if (viewElements != null) {
                    elementTargeting.setTargetElements(viewElements)
                }
            }
            else -> result.notImplemented()
        }
    }

    override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel.setMethodCallHandler(null)
    }

    internal fun registerEmbed(frameId: String, frame: AppcuesFrameView) {
        implementation.registerEmbed(frameId, frame)
    }

    private fun Result.badArgs(names: String) {
        error("bad-args", "missing one or more required args", names)
    }
}

File: ios/appcues_flutter/Sources/appcues_flutter/AppcuesFlutterPlugin.swift
import Flutter
import UIKit
import AppcuesKit

@available(*, unavailable, renamed: "AppcuesFlutterPlugin")
public typealias SwiftAppcuesFlutterPlugin = AppcuesFlutterPlugin

public class AppcuesFlutterPlugin: NSObject, FlutterPlugin {

    private static var implementation: Appcues?
    private var analyticsChannel: FlutterEventChannel?
    private var eventSink: FlutterEventSink?

    public static func register(with registrar: FlutterPluginRegistrar) {
        let methodChannel = FlutterMethodChannel(name: "appcues_flutter", binaryMessenger: registrar.messenger())
        let instance = AppcuesFlutterPlugin()
        instance.analyticsChannel = FlutterEventChannel(name: "appcues_analytics", binaryMessenger: registrar.messenger())
        registrar.addMethodCallDelegate(instance, channel: methodChannel)

        if #available(iOS 13.0, *) {
            Appcues.elementTargeting = FlutterElementTargeting()
        }

        let factory = AppcuesFrameViewFactory(plugin: instance, messenger: registrar.messenger())
        registrar.register(factory, withId: "AppcuesFrameView")
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {

        // init is a special case, which creates the Appcues instance, must be done first
        if call.method == "initialize" {
            if let accountID = call["accountId"], let applicationID = call["applicationId"] {
                let config = Appcues.Config(accountID: accountID, applicationID: applicationID)
                var enableUniversalLinks = false

                if let arguments = call.arguments as? [String: Any] {

                    if let options = arguments["options"] as? [String: Any?] {
                        if let logging = options["logging"] as? Bool {
                            config.logging(logging)
                        }

                        if let apiHost = options["apiHost"] as? String, let url = URL(string: apiHost) {
                            config.apiHost(url)
                        }

                        if let settingsHost = options["settingsHost"] as? String, let url = URL(string: settingsHost) {
                            config.settingsHost(url)
                        }

                        if let sessionTimeout = options["sessionTimeout"] as? UInt {
                            config.sessionTimeout(sessionTimeout)
                        }

                        if let activityStorageMaxSize = options["activityStorageMaxSize"] as? UInt {
                            config.activityStorageMaxSize(activityStorageMaxSize)
                        }

                        if let activityStorageMaxAge = options["activityStorageMaxAge"] as? UInt {
                            config.activityStorageMaxAge(activityStorageMaxAge)
                        }

                        enableUniversalLinks = options["enableUniversalLinks"] as? Bool ?? false
                    }

                    if let additionalAutoProperties = arguments["additionalAutoProperties"] as? [String: Any?] {
                        config.additionalAutoProperties(additionalAutoProperties.compactMapValues { $0 })
                    }
                }

                config.enableUniversalLinks(enableUniversalLinks)

                Self.implementation = Appcues(config: config)
                handleStoredNativePush()
                analyticsChannel?.setStreamHandler(self)
                result(nil)
            } else {
                result(missingArgs(names: "accountId, applicationId"))
            }
            return
        }

        guard let implementation = Self.implementation else {
            result(FlutterError(code: "notInitialized",
                                message: "the initialize function must be called before any other Appcues SDK calls",
                                details: nil))
            return
        }

        switch call.method {
        case "identify":
            if let userId = call["userId"] {
                implementation.identify(userID: userId, properties: call.properties)
                result(nil)
            } else {
                result(missingArgs(names: "userId"))
            }
        case "group":
            implementation.group(groupID: call["groupId"], properties: call.properties)
            result(nil)
        case "track":
            if let name = call["name"] {
                implementation.track(name: name, properties: call.properties)
                result(nil)
            } else {
                result(missingArgs(names: "name"))
            }
        case "screen":
            if let title = call["title"] {
                implementation.screen(title: title, properties: call.properties)
                result(nil)
            } else {
                result(missingArgs(names: "title"))
            }
        case "anonymous":
            implementation.anonymous()
            result(nil)
        case "reset":
            implementation.reset()
            result(nil)
        case "version":
            result(implementation.version())
        case "debug":
            implementation.debug()
            result(nil)
        case "show":
            if let experienceID = call["experienceId"] {
                implementation.show(experienceID: experienceID) { success, error in
                    if success {
                        result(nil)
                    } else {
                        result(FlutterError(
                            code: "show-experience-failure",
                            message: "unable to show experience \(experienceID)",
                            details: error?.localizedDescription))
                    }
                }
            } else {
                result(missingArgs(names: "experienceId"))
            }
        case "didHandleURL":
            if let urlString = call["url"], let url = URL(string: urlString) {
                result(implementation.didHandleURL(url))
            } else {
                result(missingArgs(names: "url"))
            }

        case "setTargetElements":
            guard #available(iOS 13.0, *),
                  let flutterElementTargeting = Appcues.elementTargeting as? FlutterElementTargeting,
                  let viewElements = call.parameters?["viewElements"] as? Array<Dictionary<String, Any>> else {
                return
            }
            flutterElementTargeting.setTargetElements(viewElements: viewElements)
        default:
            result(FlutterMethodNotImplemented)
        }
    }


    internal func register(frameID: String, for view: AppcuesFrameView, on parentViewController: UIViewController) {
        Self.implementation?.register(frameID: frameID, for: view, on: parentViewController)
    }

    private func missingArgs(names: String) -> FlutterError {
        return FlutterError(code: "bad-args", message: "missing one or more required args", details: names)
    }
}

extension AppcuesFlutterPlugin {
    private static var pushToken: Data?
    private static var notificationResponse: UNNotificationResponse?

    @objc
    public static func setPushToken(_ deviceToken: Data?) {
        guard let impl = Self.implementation else {
            Self.pushToken = deviceToken
            return
        }

        impl.setPushToken(deviceToken)
    }

    @objc
    public static func didReceiveNotification(response: UNNotificationResponse, completionHandler: @escaping () -> Void) -> Bool {
        guard let impl = Self.implementation else {
            Self.notificationResponse = response
            return false
        }

        return impl.didReceiveNotification(response: response, completionHandler: completionHandler)
    }

    // To be called at setup
    private func handleStoredNativePush() {
        if let pushToken = Self.pushToken {
            Self.implementation?.setPushToken(pushToken)
        }
        if let notification = Self.notificationResponse {
            _ = Self.implementation?.didReceiveNotification(response: notification, completionHandler: {})
        }
    }
}

extension AppcuesFlutterPlugin: AppcuesAnalyticsDelegate {
    public func didTrack(analytic: AppcuesKit.AppcuesAnalytic, value: String?, properties: [String : Any]?, isInternal: Bool) {
        let analyticName: String
        switch analytic {
        case .event:
            analyticName = "EVENT"
        case .screen:
            analyticName = "SCREEN"
        case .identify:
            analyticName = "IDENTIFY"
        case .group:
            analyticName = "GROUP"
        }

        eventSink?([
            "analytic": analyticName,
            "value": value ?? "",
            "properties": properties ?? [:],
            "isInternal": isInternal
        ])
    }
}

extension AppcuesFlutterPlugin: FlutterStreamHandler {
    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        eventSink = events
        Self.implementation?.analyticsDelegate = self
        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        Self.implementation?.analyticsDelegate = nil
        return nil
    }
}

private extension FlutterMethodCall {
    var parameters: Dictionary<String, Any>? {
        arguments as? Dictionary<String, Any>
    }

    var properties: Dictionary<String, Any>? {
        parameters?["properties"] as? Dictionary<String, Any>
    }

    func getParam(_ name: String) -> String? {
        guard let parameters = parameters else { return nil }
        return parameters[name] as? String
    }

    subscript(index: String) -> String? {
        get {
            getParam(index)
        }
    }
}
